mmm_experiments/agents/scientific_value.py

```diff
@@ -0,0 +1,252 @@
+        appends the proper caches.
+
+        Parameters
+        ----------
+        position : array_like
+            The position of the motor. Must match the dimensionality of that
+            provided in the ``__init__``.
+        observation : array_like
+            The observation, such as a spectrum, PDF, XRD, etc. Must match the
+            dimensionality of that provided in the ``__init__``.
+
+        Returns
+        -------
+        dict
+        """
+
+        # Using the relative position for all learning
+        relative_position = position - self.measurement_origin
+
+        self._positions_cache.append(position)
+        self._relative_positions_cache.append(relative_position)
+        self._observations_cache.append(observation)
+
+        # Compute the value
+        X = np.array(self._relative_positions_cache)
+        Y = np.array(self._observations_cache)
+        V = scientific_value_function(X, Y, sd=self._length_scale)
+
+        # The value is a scalar
+        V = V.reshape(-1, 1)
+
+        return dict(
+            position=[position],
+            rel_position=[relative_position],
+            observation=observation,
+            cache_len=[len(self._relative_positions_cache)],
+            value=V.squeeze()[-1]
+        )
+
+    def ask(self, optimize_acqf_kwargs=None):
+
+        train_x = torch.tensor(
+            self._relative_positions_cache, dtype=torch.float
+        )
+        train_x = train_x.to(self._device)
+        train_y = torch.tensor(self._observations_cache, dtype=torch.float)
+        train_y = train_y.to(self._device)
+
+        gp = SingleTaskGP(train_x, train_y).to(self.device)
+        mll = ExactMarginalLogLikelihood(gp.likelihood, gp).to(self.device)
+        fit_gpytorch_model(mll)
+
+        if optimize_acqf_kwargs is None:
+            optimize_acqf_kwargs = self._optimize_acqf_kwargs
+
+        if optimize_acqf_kwargs["q"] == 1:
+            acq = ExpectedImprovement(gp, best_f=torch.max(train_y).item())
+        else:
+            acq = qExpectedImprovement(gp, best_f=torch.max(train_y).item())
+
+        next_points, acq_value = optimize_acqf(
+            acq,
+            bounds=self.relative_bounds,
+            **optimize_acqf_kwargs,
+        )
+
+        if self._possible_coordinates is not None:
+            next_points = next_closest_raster_scan_point(
+                next_points,
+                train_x.detach().numpy(),
+                self._possible_coordinates,
+                eps=1e-8
+            )
+
+        if optimize_acqf_kwargs["q"] == 1:
+            next_points = [float(next_points.to("cpu"))]
+        else:
+            next_points = [float(x.to("cpu")) for x in next_points]
+
+        doc = dict(
+            batch_size=optimize_acqf_kwargs["q"],
+            next_points=next_points,
+            acq_value=[float(x.to("cpu")) for x in acq_value]
+            if optimize_acqf_kwargs["q"] > 1
+            else [float(acq_value.to("cpu"))],
+        )
```

Comment:
```suggestion
        doc = dict(
            batch_size=[optimize_acqf_kwargs["q"]],
            next_points=[next_points],
            acq_value=[float(x.to("cpu")) for x in acq_value]
            if optimize_acqf_kwargs["q"] > 1
            else [float(acq_value.to("cpu"))],
        )
```
For the way we are currently publishing docs, even lists belong inside lists. Everything is an "event page", which is a collection of "events". A hack that will not be permanent but exists for now.